### PR TITLE
[flang][OpenMP] Delayed privatization for variables with `equivalence` association

### PR DESCRIPTION
**Pull request by ergawy**
Handles variables that are storage associated via `equivalence`. The problem is that these variables are declared as `fir.ptr`s while their privatized storage is declared as `fir.ref` which was triggering a validation error in the OpenMP dialect.

### CODE REVIEW

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (Dinistro) [inline mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:2508]: That might be doable by working with dialect interfaces, I think.

**Review by tblah — COMMENTED** [review]
LGTM, but please wait for another approval because I am still getting up to speed with delayed privatisation.

File: mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp
Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:2508]: Probably, but I couldn't find any interfaces declared by FIR types that can help here. @kiranchandramohan is there a way to do that right now?

To clarify, given 2 types, for example: `fir.ptr<i32>` and `fir.ref<i32>`, I want to test whether they are pointer/reference-like types that point/refer to the same underlying type.

Comment (Dinistro) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:2508]: I'm talking about dialect interfaces, not type interfaces (similar to the `ConvertToLLVMPatternInterface`). These could be used to implement arbitrary dialect specific logic that this verifier could delegate to. Note that this might be a bit of overkill, though.

Comment (kiranchandramohan) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp:2508]: Can you add suitable `fir.convert` operations to enforce the matching type in the generated code?

Comment (ergawy) [reply mlir/lib/Dialect/OpenMP/IR/OpenMPDialect.cpp]: Done! Thanks for the suggestion.

**Comment by ergawy** [comment]
@kiranchandramohan @Dinistro Thanks for the review, do you have any further comments/suggestions on this PR?

**Comment by Dinistro** [comment]
> @kiranchandramohan @Dinistro Thanks for the review, do you have any further comments/suggestions on this PR?

I'm fine with the implemented solution and do not have the expertise to properly review the flang parts. If this looks good to others, it's good for me 🙂

File: mlir/test/Target/LLVMIR/openmp-private.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/openmp-private.mlir:268]:
```suggestion
// Test that we initialize the firstprivate variable.
```

File: mlir/test/Target/LLVMIR/openmp-private.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/openmp-private.mlir:270]: Nit: Shouldn't there be a `CHECK-NOT: define {{.*}} @{{.*}}` or similar line to make sure the `store` instruction below isn't part of another function?

**Review by skatrak — APPROVED** [review]
Thank you Kareem, this LGTM. However, I'm not very familiar with `equivalence` handling in flang, so perhaps it's still good waiting for a look by another reviewer.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:561]: Nit: Can you add a comment on where the convert is usually required?

**Review by kiranchandramohan — APPROVED** [review]
LG.

**Pull request merged**
